src/analysis/inner_product_space/lax_milgram.lean

```diff
@@ -0,0 +1,147 @@
+
+lemma injective (coercive : is_coercive B) : (lax_milgram_map B).ker = ⊥ :=
+begin
+  simp only [submodule.eq_bot_iff, continuous_linear_map.mem_ker],
+  intros v,
+  contrapose,
+  simp_rw [← ne.def, ← norm_pos_iff],
+  intros hv,
+  rcases coercive with ⟨C, C_ge_0, coercivity⟩,
+  have squared : C * ∥v∥ * ∥v∥ ≤ ∥lax_milgram_map B v ∥ * ∥v∥ :=
+  calc C * ∥v∥*∥v∥
+       ≤ B v v                         : coercivity v
+  ...  = inner (lax_milgram_map B v) v : by simp
+  ...  ≤ ∥lax_milgram_map B v ∥ * ∥v∥    : real_inner_le_norm (lax_milgram_map B v) v,
+  have coerced := (mul_le_mul_right hv).mp squared,
+  exact calc 0 < C * ∥v∥                  : mul_pos C_ge_0 hv
+  ...          ≤ ∥(lax_milgram_map B) v∥  : coerced,
+end
+
+
+lemma closed_range (coercive : is_coercive B) : is_closed ((lax_milgram_map B).range : set V) :=
+begin
+  rw ←is_seq_closed_iff_is_closed,
+  apply is_seq_closed_of_def,
+  intros x y mem_range_x tendsto_y,
+  sorry,
+end
+
+lemma surjective (coercive : is_coercive B): (lax_milgram_map B).range = ⊤ :=
+begin
+  rw submodule.eq_top_iff', intro v,
+  -- As the range of lax_milgram_map is closed, it is also complete
+  haveI := (closed_range coercive).complete_space_coe,
+  -- Every vector can be decomposed uniquely as the sum of an element in the range
+  -- and an element in the complement
+  have decompose_v :=
+    eq_sum_orthogonal_projection_self_orthogonal_complement (lax_milgram_map B).range v,
+  -- We show the orthogonal component of v is 0
+  have zero_orthogonal_component
+    : ↑((orthogonal_projection (lax_milgram_map B).rangeᗮ) v) = (0 : V),
+  by begin
+    -- the orthogonal component of v is zero iff, for all w in the orthogonal complement ⟪v,w⟫=0
+    refine eq_orthogonal_projection_of_mem_of_inner_eq_zero
+      (submodule.zero_mem ((lax_milgram_map B).range)ᗮ) _,
+    intros w mem_w_orthogonal,
+    -- We show every element of the orthogonal complement is zero, by coercivity
+    rcases coercive with ⟨C, C_ge_0, coercivity⟩,
+    have nonneg_C : 0 ≤ C, by linarith,
+    have : 0 ≤ C * ∥w∥*∥w∥ := mul_nonneg (mul_nonneg nonneg_C (norm_nonneg w)) (norm_nonneg w),
+    have inner_product_eq_zero :=
+      (submodule.mem_orthogonal ((lax_milgram_map B).range) w).mp mem_w_orthogonal,
+    have : lax_milgram_map B w ∈ (lax_milgram_map B).range :=
+      by simp [continuous_linear_map.mem_range],
+    have : C * ∥w∥*∥w∥ ≤ 0 :=
+    calc C * ∥w∥*∥w∥
+         ≤ B w w                          : coercivity w
+    ...  = inner (lax_milgram_map B w) w  : by simp
+    ...  = 0                              : by {apply inner_product_eq_zero, assumption},
+    have zero_eq_C_norm_w_squared : C * ∥w∥*∥w∥ = 0 := by linarith,
+    have w_eq_zero : w = 0 := by begin
+      simp at zero_eq_C_norm_w_squared,
+      cases zero_eq_C_norm_w_squared, -- Either C or w is zero
+       -- if C = 0 we get a contradiction
+      exfalso, linarith,
+      -- if w = 0 we've shown what we wanted
+      assumption,
+    end,
```

Comment:
And for this part, it's still more painful than one would wish, but you can get rid of this and lines 112-113 with the following endgame:
```lean
  have : ∥w∥ * ∥w∥ ≤ 0 := by nlinarith,
  have h : ∥w∥ = 0 := by nlinarith [norm_nonneg w],
  have w_eq_zero : w = 0 := by simpa using h,
  simp [w_eq_zero],
```